### ISSUE TITLE
iOS 15 is often disconnected when I go to use the tailnet

### ISSUE THREAD

**Issue by offbyone** [tailscale/tailscale#4646]
### What is the issue?

This is an iOS 15 clone of #3187 because I accidentally misreported my iOS version in there, confusing it with my  phone model number (I am somewhat smart, I swear!)

When I attempt to connect to my tailnet, after a long period not using it, it is disconnected.

I'm not sure how to generate a bugreport from the tailscale iOS application, so I'm including one from my macOS node, *but that machine is not experiencing the issue*.

### Steps to reproduce

Attempt to connect to an IP on my tailnet, and it is not connected. If I open the tailscale app, then try again, it works.

### Are there any recent changes that introduced the issue?

No

### OS

iOS

### OS version

15.4.1

### Tailscale version

1.24.2

### Bug report

BUG-42b3fcc4c8eda420cb4e6deea027750e053580e35b79bb91f2c3c1a27b3d971d-20220509053951Z-3811e5dcadd2b3eb

**Comment by maisem** [comment]
The only time I've noticed this is when my phone has restarted, e.g. after an OS upgrade. Do you remember if your phone had also restarted recently?

**Comment by offbyone** [comment]
I have not restarted it recently that I can recall.

**Comment by bswinnerton** [comment]
Experiencing this as well. I tried upgrading to the beta (1.26), as well as iOS 15.5, neither of which resolved the problem. 

I can't seem to determine what causes this, other than time. It's not a state change between wifi and LTE, and it doesn't seem to happen when turning the screen off or force closing Tailscale.

**Comment by digitaldan** [comment]
I can confirm this as well.  After some amount of time, say overnight , tailscale is no longer connected,  my IOS device shows the the "VPN" is not  active in the IOS system settings, and there's no "VPN" status icon on the top of then screen.  If i then go to launch  the tailscale app,  it immediately attempts to connect and succeeds, but eventually will die again some amount of hours later.  My phone is not restarting during these times.   I don't think i have seen more than 24 hours being connected.   I would be happy to post logs if i knew where they were :-)

**Comment by DoFoT9** [comment]
I have also been experiencing this in iOS15 (and iPadOS 15). I have upgraded to iOS16/iPadOS16 earlier this week, and it certainly seems to retain the connection for a longer period - currently at 15 hours. I believe this has to do with the way that memory works on your iDevices. My temporary workaround is to open the app on my devices every few hours.

If there is a permeant fix, I would greatly appreciate hearing it!

**Comment by bswinnerton** [comment]
Turning off MagicDNS has had success for me so far. No disconnects in the last 24 hours.

**Comment by digitaldan** [comment]
FYi, this seems to be a related or a DUP of https://github.com/tailscale/tailscale/issues/337 .  That ticket mentions to enable "connect on demand" in the IOS VPN settings, but that is not available for me in my settings menu, not sure if Apple removed this, or if this is a tailscale option that would need to be enabled.

**Comment by DoFoT9** [comment]
@digitaldan - I believe the settings are managed via VPN profiles/etc, so the "always on" setting is controlled there. I am unfortunately still experiencing the same problems on iOS16/iPadOS16 beta 2 (released June 21/2022). Interestingly, I can't seem to be able to find any TailScale logs on my phone to assist in the diagnosis process.

**Comment by digitaldan** [comment]
So this just happened again after being up for 2 or more days, vpn is not connected and phone has not been rebooted in 7 days.  Additionally there is no log file in the analytics section in the privacy settings menu .  If I go into the Tailscale app, it immediately connects without prompting for authorization.

**Comment by DoFoT9** [comment]
> So this just happened again after being up for 2 or more days, vpn is not connected and phone has not been rebooted in 7 days. Additionally there is no log file in the analytics section in the privacy settings menu . If I go into the Tailscale app, it immediately connects without prompting for authorization.

I've still got all 3 of my iOS 15/16 devices randomly cutting out. The app reopens, so I'm guessing it is thrown out of active memory.

**Comment by nicholasgcotton** [comment]
Ditto, same issue here. Happens daily on my iPhone 11 Pro and iPad Pro M1. I leave my home WiFi and have no data on transit until I cycle tailscale off and on. 
BUG-8da0a06b988417b9a1248065eb22897188509d002a1066f4edd0ed5695e78fe5-20220808213719Z-968e4f93ba98d0b6

**Comment by digitaldan** [comment]
Yeah, i'm surprised this is not annoying Tailscale developers and staff, i assume this is happening on most people's iPhones?  Or is this only for us select lucky few ?  It's really crippling the usefulness of this solution for us, it continues to be an issue on my devices. 

I see this is tagged as a "L2 Few", but is that really the case?  Is this not happening to nearly all IOS users ?  I  really wish i could help debug, but there is no crash log when this happens.

**Comment by offbyone** [comment]
This is persisting on the iOS 16 public betas.

**Comment by digitaldan** [comment]
So I have started recording the "last seen" time when my phone disconnects. So for example if i notice tailscale is not running, i go to the admin console and record the "last seen" time before connecting my phone again.  The results definitely show a predictable pattern.  I have not caught all of them, but most over the last 2 weeks. 

To set a little background, my phone is a iPhone 13 Max, nothing weird, no MDM or other supervisory software installed.  Before i go to bed around 11pm, i put my phone on a magsafe charger.  The phone does have the default "Optimized Battery Charging" turned on.  I only mention this as after seeing the pattern myself, my first question was what would the phone be doing at night around the same time to cause a disconnect?

Aug 30 1:39 AM
Sep 6 1:55 AM
Sep 14, 2:01 AM
Sep 15, 2:40 AM
Sep 16, 3:01 AM

So its happening not only at night, but around the same time.  In fact its slightly skewing later over time.  I have also confirmed that my phone did not reboot during these disconnects.  FYI i upgraded to IOS 16 when it WENT GA, so the problem persists on newer software as well. 

Hope this helps and that someone is actively taking an interest in this issue!

**Comment by DoFoT9** [comment]
Thanks to all for their insights and support, very interesting to note that I'm not the only one!

In addition to the above, it seems that the toggle switch to turn on/off the VPN is not responding as expected, and in fact some times ignores your input completely, but I guess that's another ticket.

I am still experiencing drops every day or so, even though it is always left on. I must routinely open the app to 'remind' iOS/iPadOS that there is a VPN it needs to retain. Frustrating.

**Comment by digitaldan** [comment]
Well I think i found something.

While debugging another app, I was searching the log files of my phone and found a few "SystemMemoryReset.." logs under the phones "Settings" -> "Privacy & Security" -> "Analytics & Improvements" -> "Analytic Data" .  These logs are produced when the system crosses a memory threshold and will trigger a restart of the OS. 

The timestamp of the log (which i will post below) matches up exactly to the times my phone was "last seen" in the tailscale admin console. It seems like my phone only keeps the last 3 logs FYI.  Whats interesting is that i don't think this is a hard hardware reboot, i use iSH for very light console access, and the busy box "uptime" command shows me the correct uptime  since the last "hard" reboot on my phone (so manually rebooting or after a software update). 

So this indeed comes back to https://github.com/tailscale/tailscale/issues/3771 , having the application launch at boot seems pretty critical at this point.

here is a snippet of the log file from 9/24 at 1:29 am PST, the admin panel showed the last seen time as "Sep 24, 1:29 AM" (see my earlier post) 

```json
{"bug_type":"301","timestamp":"2022-09-24 01:29:23.00 -0700","os_version":"iPhone OS 16.0 (20A362)","roots_installed":0,"incident_id":"EFD096B3-B570-49E9-855E-7CFD1A90F523"}
{
  "crashReporterKey" : "ff7a2949049ea22beb21e76fbf575d7d65cfda2d",
  "kernel" : "Darwin Kernel Version 22.0.0: Tue Aug 16 20:51:28 PDT 2022; root:xnu-8792.2.11.0.1~1\/RELEASE_ARM64_T8110",
  "product" : "iPhone14,3",
  "incident" : "EFD096B3-B570-49E9-855E-7CFD1A90F523",
  "date" : "2022-09-24 01:29:23.80 -0700",
  "build" : "iPhone OS 16.0 (20A362)",
  "roots_installed" : 0,
  "bug_type" : "301",
  "eventCode" : 0,
  "eventReason" : "User reclaimable memory dropped below the limit. User reclaimable current: 69%. User reclaimable minimum: 70%",
  "timeDelta" : 2094757701,
  "memoryStatus" : {
  "compressorSize" : 60743,
  "compressions" : 22433172,
  "decompressions" : 17270300,
  "zoneMapCap" : 2160738304,
  "largestZone" : "APFS_4K_OBJS",
  "largestZoneSize" : 41074688,
  "pageSize" : 16384,
  "uncompressed" : 195032,
  "zoneMapSize" : 287113216,
  "memoryPages" : {
    "active" : 112968,
    "throttled" : 0,
    "fileBacked" : 109194,
    "wired" : 102080,
    "anonymous" : 116363,
    "purgeable" : 4365,
    "inactive" : 108618,
    "free" : 5187,
    "speculative" : 3971
  }
}
```

**Comment by digitaldan** [comment]
FYI, Googling a little bit on this points to IOS doing a "user space" reboot vs a full system reboot.

**Comment by shrop** [comment]
I don't think I had this issue before iPhone 14 Pro Max with iOS 16.0.1. Seems to happen overnight while charging. Able to open the Tailscale app in the AM and reconnect no problem.

**Comment by DoFoT9** [comment]
Ditto - similar experience for me overnight last night.

**Comment by jronnen7** [comment]
I have experienced a similar issue, except it appeared connected and the requests were simply timing out.  I say this because I could disconnect from the VPN (in app) and receive a 403 error from the box (as I expected).  But, as soon as I connected in app it would timeout and never server the page.

The only way to fix it was going into the settings menu and toggling VPN.  Reinstalling the app didn't fix it.  IOS 15.7

**Comment by bexelbie** [comment]
It appears I am facing the "disconnect because of system memory reset" problem as well.  Commenting to add a +1 and to follow.

**Comment by sp00ls** [comment]
I'm also experiencing this issue on both iOS 16 & the iOS 17 developer beta. When I try to use an app that connects to a device on my tailnet after a period of inactivity it hangs & no connection can be established. Opening the Tailscale app for just a moment before switching back to the other app fixes it.

**Comment by DoFoT9** [comment]
Hi all - still experiencing this issue, so I have submitted my logs under iOS 17 beta 3 using the "Feedback" app that comes by default on every beta OS. I highly recommend that all parties effected by this also contribute in a similar manner so the kind folks at Apple can get as much info as possible to find a resolution (or to identify where the problem lies). 

Watch this space

DoFo.

**Comment by digitaldan** [comment]
> so the kind folks at Apple can get as much info as possible to find a resolution 

Well, the original posted issue on this bug is not an Apple issue but a tailscale issue.  IOS apps are expected to handle their lifecycle properly, which includes being disconnected, backgrounded or terminated.  The Tailscale IOS app is not taking advantage of the proper IOS APIs which would allow the app to reconnect.  Other apps, like the open source Wireguard client, on which the Tailscale protocol is based, has no problems reconnecting.  I am currently using this on IOS beta with no issues as my primary VPN client now b/c of this particular bug.

**Comment by digitaldan** [comment]
So i see that a redesigned IOS app is out ,  i'm curious if this issue has been resolved?  I may run it for a few days and see.

**Comment by DoFoT9** [comment]
> So i see that a redesigned IOS app is out , i'm curious if this issue has been resolved? I may run it for a few days and see.
> 
> https://tailscale.com/blog/reimagining-tailscale-for-ios/

Hi Dan - I’ve just installed the latest betas (via TestFlight) onto my various machines and will be monitoring over the coming days. Will report back with any updates, I have not checked to see if there are any notes specifically addressing this.

Thanks,
DoFo.

**Comment by DoFoT9** [comment]
Hi All - I continue to have timeouts/app lockups overnight, so it appears that this has not been addressed in the latest round of changes. 

Thanks,
DoFo

**Comment by digitaldan** [comment]
Yeah, after a day or two i noticed the connection had died and did not reconnect.  Sticking to wireguard until this is resolved, although at this point i'm not sure i'm ever coming back to tailscale  It's a shame such a critical issue can not be resolved by the Tailscale team.

**Comment by digitaldan** [comment]
This may be fixed in recent test flight builds 🤞 

See https://github.com/tailscale/tailscale/issues/1534

**Comment by digitaldan** [comment]
Also why did someone mark my comment from this week as a duplicate?  That seems pretty disrespectful. 

<img width="872" alt="image" src="https://github.com/tailscale/tailscale/assets/1903737/da0a3eb2-f111-457a-82b4-d22cd67fb6c8">

**Comment by agottardo** [comment]
I have submitted a new build to TestFlight (v1.47.71). When it lands, provided you have enabled VPN On-Demand, this issue should be fixed. The system will kill the VPN overnight as part of SystemMemoryReset, but the on-demand feature will now restart the tunnel. 

Please note this is **not** fixed in the first build with on-demand support (v1.47.67), so make sure you're using the latest version when testing.

**Comment by digitaldan** [comment]
Thanks @agottardo , thats really great news !

**Comment by DoFoT9** [comment]
Hi all - initial testing from my devices appears to be working, which is great! I have had a few unrelated app crashes which are mostly due to running iOS betas. 

I appreciate this might not be the right channel for requesting features, however, it would be great to see an option to expand the VPN on demand feature to selecting turn on/off an exit node. I like to turn on an exit node when I leave my home WiFi for example so I’m protected. 

Thanks for your hard work to all. It’s very much appreciated! 

DoFo

**Comment by bradfitz** [comment]
@digitaldan, I moderate a lot of our GitHub issues comments to keep the threads productive and on topic. You can email me (bradfitz@) if you'd like to discuss, but not in this thread.

**Comment by digitaldan** [comment]
> @digitaldan, I moderate a lot of our GitHub issues comments to keep the threads productive and on topic. You can email me (bradfitz@) if you'd like to discuss, but not in this thread.

Nothing to discuss , i think its clear to anyone who reads this thread.  I'm definitely surprised, and disappointed.

**Comment by bradfitz** [comment]
@digitaldan, you might find https://github.com/bradfitz/issue-tracker-behaviors interesting. It's important to keep the signal-to-noise ratio high to prevent people who are just monitoring this issue from unsubscribing. Those people might have had interesting things to add at some point.

So, please, let's keep this on topic. If anybody has new information to share that will be useful for debugging, please share.

**Comment by digitaldan** [comment]
> @digitaldan, you might find https://github.com/bradfitz/issue-tracker-behaviors interesting. It's important to keep the signal-to-noise ratio high to prevent people who are just monitoring this issue from unsubscribing. Those people might have had interesting things to add at some point.
> 
> So, please, let's keep this on topic. If anybody has new information to share that will be useful for debugging, please share.

I was replying to the thread with my findings, that the newly releases IOS was not reconnecting, and because of that i was sticking with an alternative solution.   

I was also the one to point out here that as new version was released and might be a solution to this issue (which we soon found out it would be in recent test flight builds) , so I think that counts as pretty relative, and not a "duplicate",  or "off-topic" which you now just changed it to.  I also think if you bothered to read my other comments, on this issue and related ones, i have been an objective and positively contributing voice to the project.  

As someone who is very involved with open source projects, i find your moderation at odds with the community which i think Tailscale wants to be associated with.  Again, I'm very disappointed in how this conversation has evolved. 

You of course have the power to moderate, and suppress what speech you find inconvenient.  I do not have that power.

**Comment by agottardo** [comment]
Closing because this can be addressed by using VPN On Demand, now available in v1.48.
